extensions-core/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/CachingKubernetesPeonClient.java

```diff
@@ -0,0 +1,284 @@
+    // This is to prevent us from waiting for entire max job runtime on a job that was canceled before it even started.
+    long jobMustBeSeenBy = startTime + (clientApi.getInformerResyncPeriodMillis() * 2);
+    boolean jobSeenInCache = false;
+
+    // Set up to watch for job changes
+    CompletableFuture<Job> jobFuture = clientApi.getEventNotifier().waitForJobChange(taskId.getK8sJobName());
+
+    // We will loop until the full timeout is reached if the job is seen in cache. If the job does not show up in the cache we will exit earlier.
+    // In this loop we first check the cache to see if our job is there and complete. This avoids missing notifications that happened before we set up the watch.
+    // If the job is not complete we wait for a notification of a job change or a timeout.
+    // If it is a timeout, we loop back to check the cache again.
+    // If it is a job change notification, we check the job state and exit if complete, or loop again if still running.
+    do {
+      try {
+        Optional<Job> maybeJob = getPeonJob(taskId.getK8sJobName());
+        if (maybeJob.isPresent()) {
+          jobSeenInCache = true;
+          Job job = maybeJob.get();
+          JobResponse currentResponse = determineJobResponse(job);
+          if (currentResponse.getPhase() != PeonPhase.RUNNING) {
+            return currentResponse;
+          } else {
+            log.debug("K8s job[%s] found in cache and is still running", taskId.getK8sJobName());
+          }
+        } else if (jobSeenInCache) {
+          // Job was in cache before, but now it's gone - it was deleted and will never complete.
+          log.warn("K8s Job[%s] was not found. It can happen if the task was canceled", taskId.getK8sJobName());
+          return new JobResponse(null, PeonPhase.FAILED);
+        }
+
+        // We wake up every informer resync period to avoid event notifier misses.
+        Job job = jobFuture.get(clientApi.getInformerResyncPeriodMillis(), TimeUnit.MILLISECONDS);
+
+        // Immediately set up to watch for the next change in case we need to wait again
+        jobFuture = clientApi.getEventNotifier().waitForJobChange(taskId.getK8sJobName());
+        log.debug("Received job[%s] change notification", taskId.getK8sJobName());
+        jobSeenInCache = true;
+
+        if (job == null) {
+          log.warn("K8s job[%s] was not found. It can happen if the task was canceled", taskId.getK8sJobName());
+          return new JobResponse(null, PeonPhase.FAILED);
+        }
+
+        JobResponse currentResponse = determineJobResponse(job);
+        if (currentResponse.getPhase() != PeonPhase.RUNNING) {
+          return currentResponse;
+        } else {
+          log.debug("K8s job[%s] is still running", taskId.getK8sJobName());
+        }
+      }
+      catch (TimeoutException e) {
+        // A timeout here is not a problem, it forces us to loop around and check the cache again.
```

Comment:
If the future times out, do we need to remove it from the event notifier map proactively?
I think it would get cleaned up only if we receive an event for that job/pod. Otherwise, it will always remain in memory.

I am not sure if it would be safe to just de-register using a `stopWatchingJobChange()` method, as it would essentially mark (all) the futures for that job as cancelled, thus causing exception in any other thread waiting for the future to get resolved.

I guess to handle that case by catching a `CancellationException` and just looping back.

@capistrant , what do you think?

Reply:
Ya there is definitely a problem here. I am not positive on how to fix it yet. I am working on one angle that uses an atomic counter associated with each future so something that is not referenced anymore can be cleaned up proactively. But I'm not so sure that is going to work either. I'm going to keep thinking about this tonight and try to come up with a legitimate fix. The pattern just might be too brittle. Will also think about your idea of tolerating a cancelation exception and looping back around.

Reply:
Thanks, @capistrant!

The reference counting approach could be a viable option. `HeapMemoryDatasourceSegmentCache` does something similar.